pkg/controller/disruption/disruption.go

```diff
@@ -0,0 +1,456 @@
+			currentHealthy++
+		}
+	}
+	return
+}
+
+func (r *ReconcileMachineDisruption) updateMachineDisruptionBudgetStatus(
+	mdb *healthcheckingv1alpha1.MachineDisruptionBudget,
+	currentHealthy,
+	desiredHealthy,
+	expectedCount int32,
+	disruptedMachines map[string]metav1.Time) error {
+
+	// We require expectedCount to be > 0 so that MDBs which currently match no
+	// machines are in a safe state when their first machines appear but this controller
+	// has not updated their status yet.  This isn't the only race, but it's a
+	// common one that's easy to detect.
+	disruptionsAllowed := currentHealthy - desiredHealthy
+	if expectedCount <= 0 || disruptionsAllowed <= 0 {
+		disruptionsAllowed = 0
+	}
+
+	if mdb.Status.CurrentHealthy == currentHealthy &&
+		mdb.Status.DesiredHealthy == desiredHealthy &&
+		mdb.Status.ExpectedMachines == expectedCount &&
+		mdb.Status.MachineDisruptionsAllowed == disruptionsAllowed &&
+		reflect.DeepEqual(mdb.Status.DisruptedMachines, disruptedMachines) &&
+		mdb.Status.ObservedGeneration == mdb.Generation {
+		return nil
+	}
+
+	newMdb := mdb.DeepCopy()
+	newMdb.Status = healthcheckingv1alpha1.MachineDisruptionBudgetStatus{
+		CurrentHealthy:            currentHealthy,
+		DesiredHealthy:            desiredHealthy,
+		ExpectedMachines:          expectedCount,
+		MachineDisruptionsAllowed: disruptionsAllowed,
+		DisruptedMachines:         disruptedMachines,
+		ObservedGeneration:        mdb.Generation,
+	}
+
+	return r.client.Update(context.TODO(), newMdb)
+}
+
+// failSafe is an attempt to at least update the MachineDisruptionsAllowed field to
+// 0 if everything else has failed.  This is one place we
+// implement the  "fail open" part of the design since if we manage to update
+// this field correctly, we will prevent the deletion when it may be unsafe to do
+func (r *ReconcileMachineDisruption) failSafe(mdb *healthcheckingv1alpha1.MachineDisruptionBudget) error {
+	newMdb := mdb.DeepCopy()
+	mdb.Status.MachineDisruptionsAllowed = 0
+	return r.client.Update(context.TODO(), newMdb)
+}
+
+func (r *ReconcileMachineDisruption) getMachineDisruptionBudgetForMachine(machine *mapiv1.Machine) *healthcheckingv1alpha1.MachineDisruptionBudget {
+	// GetMachineMachineDisruptionBudgets returns an error only if no
+	// MachineDisruptionBudgets are found.  We don't return that as an error to the
+	// caller.
+	mdbs, err := machineutil.GetMachineMachineDisruptionBudgets(r.client, machine)
+	if err != nil {
+		glog.V(4).Infof("No MachineDisruptionBudgets found for machine %v, MachineDisruptionBudget controller will avoid syncing.", machine.Name)
+		return nil
+	}
+
+	if len(mdbs) == 0 {
+		glog.V(4).Infof("Could not find MachineDisruptionBudget for machine %s in namespace %s with labels: %v", machine.Name, machine.Namespace, machine.Labels)
+		return nil
+	}
+
+	if len(mdbs) > 1 {
+		msg := fmt.Sprintf("Machine %q/%q matches multiple MachineDisruptionBudgets.  Chose %q arbitrarily.", machine.Namespace, machine.Name, mdbs[0].Name)
```

Comment:
Does this mean the be other owing mdb won't never be updated?

Reply:
It the same logic that PDB has, we can specify it under the Documentation.